packages/backend/migrations/20240809171654_add_ilp_quote_details_table.js

```diff
@@ -0,0 +1,130 @@
+          table.dropColumn('maxPacketAmount')
+          table.dropColumn('minExchangeRateNumerator')
+          table.dropColumn('minExchangeRateDenominator')
+          table.dropColumn('lowEstimatedExchangeRateNumerator')
+          table.dropColumn('lowEstimatedExchangeRateDenominator')
+          table.dropColumn('highEstimatedExchangeRateNumerator')
+          table.dropColumn('highEstimatedExchangeRateDenominator')
+        })
+      })
+  )
+}
+
+/**
+ * @param { import("knex").Knex } knex
+ * @returns { Promise<void> }
+ */
+exports.down = function (knex) {
+  return knex.schema
+    .alterTable('quotes', function (table) {
+      // restore columns without not null constraint
+      table.bigInteger('maxPacketAmount')
+      table.decimal('minExchangeRateNumerator', 64, 0)
+      table.decimal('minExchangeRateDenominator', 64, 0)
+      table.decimal('lowEstimatedExchangeRateNumerator', 64, 0)
+      table.decimal('lowEstimatedExchangeRateDenominator', 64, 0)
+      table.decimal('highEstimatedExchangeRateNumerator', 64, 0)
+      table.decimal('highEstimatedExchangeRateDenominator', 64, 0)
+    })
+    .then(() => {
+      // Migrate data back to quotes table from ilpQuote
+      return knex.raw(`
+        UPDATE "quotes"
+        SET 
+          "maxPacketAmount" = "ilpQuoteDetails"."maxPacketAmount",
+          "minExchangeRateNumerator" = "ilpQuoteDetails"."minExchangeRateNumerator",
+          "minExchangeRateDenominator" = "ilpQuoteDetails"."minExchangeRateDenominator",
+          "lowEstimatedExchangeRateNumerator" = "ilpQuoteDetails"."lowEstimatedExchangeRateNumerator",
+          "lowEstimatedExchangeRateDenominator" = "ilpQuoteDetails"."lowEstimatedExchangeRateDenominator",
+          "highEstimatedExchangeRateNumerator" = "ilpQuoteDetails"."highEstimatedExchangeRateNumerator",
+          "highEstimatedExchangeRateDenominator" = "ilpQuoteDetails"."highEstimatedExchangeRateDenominator"
+        FROM "ilpQuoteDetails"
+        WHERE "quotes"."id" = "ilpQuoteDetails"."quoteId"
+      `)
+    })
+    .then(() => {
+      // Apply the not null constraints after data insertion
+      return knex.schema.alterTable('quotes', function (table) {
+        table.bigInteger('maxPacketAmount').notNullable().alter()
+        table.decimal('minExchangeRateNumerator', 64, 0).notNullable().alter()
+        table.decimal('minExchangeRateDenominator', 64, 0).notNullable().alter()
+        table
+          .decimal('lowEstimatedExchangeRateNumerator', 64, 0)
+          .notNullable()
+          .alter()
+        table
+          .decimal('lowEstimatedExchangeRateDenominator', 64, 0)
+          .notNullable()
+          .alter()
+        table
+          .decimal('highEstimatedExchangeRateNumerator', 64, 0)
+          .notNullable()
+          .alter()
+        table
+          .decimal('highEstimatedExchangeRateDenominator', 64, 0)
+          .notNullable()
+          .alter()
+      })
+    })
```

Comment:
this down will fail if local quotes are added before migration is run because they dont have these non null columns. We are setting back to null because that was the original state before the migration. What should we do? 
- leave them as nullable in the down. Doesnt fully restore the original state but not sure we can, short of losing the new quotes that dont fit in the previous model.
-  keep nullable and set to sensible(ish) defaults (derive from exchange rate)?

Reply:
> Let's split this up into multiple migrations, something like:
> 
> * First one will backfill `estimatedExchangeRate`, and mark the field as required
> * second one will create `ilpQuoteDetails`
> * third one will backfill `ilpQuoteDetails`
> 
> Then we deploy the code changes to start reading from `ilpQuoteDetails`
> 
> * last one will drop ilp fields from `quotes`
> 
> this way, we dont run into the risk of losing data

I'm not necessarily against splitting into multiple migrations. I suppose it gives more control over up/down if needed without downside.

In terms of this part:

> - third one will backfill ilpQuoteDetails
>
> Then we deploy the code changes to start reading from ilpQuoteDetails
>
> - last one will drop ilp fields from quotes

Is the idea that we put the drop ilp fields from quotes in another release and communicate as much in patch notes? I mean I dont see a way to ensure integrators arent just upgrading multiple releases and running them all at once in that case.

Reply:
I think this all will go in one release, I was just thinking of how to make the changes safely (or in separate PRs). 

If you want, we can have this:

> Then we deploy the code changes to start reading from ilpQuoteDetails

released the same time as we backfill `ilpQuoteDetails` (and `drop ilp fields from quotes` for that matter)

Reply:
thinking about it more, i dont think we can split the create ilpQuoteDetails table and the migration to backfill it.

if migrations are run through creatye ilpQuoteDetails (but not backill), and then quotes are created in the application with ilpQuoteDetails, then they are backfilled, then the quotes created via the application will try to be inserted again in the ilpQuoteDetails (and it will fail because the quoteId has a unique constraint).